src/generic.jl

```diff
@@ -26,49 +26,115 @@ evaluate(dist::PreMetric, a, b) = dist(a, b)
+`length(a) == length(b)`.
+"""
+function colwise!(r::AbstractArray, metric::PreMetric, a, b)
+    require_one_based_indexing(r)
+    n = length(a)
+    length(b) == n || throw(DimensionMismatch("iterators have different lengths"))
+    length(r) == n || throw(DimensionMismatch("Incorrect size of r."))
+    @inbounds for (j, ab) in enumerate(zip(a, b))
+        r[j] = metric(ab...)
+    end
+    r
+end
+
 function colwise!(r::AbstractArray, metric::PreMetric, a::AbstractVector, b::AbstractMatrix)
+    require_one_based_indexing(r)
     n = size(b, 2)
     length(r) == n || throw(DimensionMismatch("Incorrect size of r."))
-    @inbounds for j = 1:n
-        r[j] = metric(a, view(b, :, j))
+    @inbounds for (rj, bj) in enumerate(axes(b, 2))
+        r[rj] = metric(a, view(b, :, bj))
     end
     r
 end
 
 function colwise!(r::AbstractArray, metric::PreMetric, a::AbstractMatrix, b::AbstractVector)
+    require_one_based_indexing(r)
     n = size(a, 2)
     length(r) == n || throw(DimensionMismatch("Incorrect size of r."))
-    @inbounds for j = 1:n
-        r[j] = metric(view(a, :, j), b)
+    @inbounds for (rj, aj) in enumerate(axes(a, 2))
+        r[rj] = metric(view(a, :, aj), b)
     end
     r
 end
 
+"""
+    colwise!(r::AbstractMatrix, metric::PreMetric,
+              a::AbstractMatrix, b::AbstractMatrix)
+
+Compute distances between each corresponding columns of `a` and `b` according
+to distance `metric`, and store the result in `r`. Exactly one of `a` or `b`
+can be a vector, in which case the distance between that vector and all columns
+of the other matrix are computed.
+
+`a` and `b` must have the same number of columns if neither of the two is a
+vector. `r` must be a vector of length `maximum(size(a, 2), size(b, 2))`.
+"""
 function colwise!(r::AbstractArray, metric::PreMetric, a::AbstractMatrix, b::AbstractMatrix)
+    require_one_based_indexing(r, a, b)
     n = get_common_ncols(a, b)
     length(r) == n || throw(DimensionMismatch("Incorrect size of r."))
-    @inbounds for j = 1:n
+    @inbounds for j in 1:n
         r[j] = metric(view(a, :, j), view(b, :, j))
     end
     r
 end
 
-function colwise!(r::AbstractArray, metric::SemiMetric, a::AbstractMatrix, b::AbstractVector)
-    colwise!(r, metric, b, a)
+"""
+    colwise(r::AbstractMatrix, metric::PreMetric, a, b)
+
+Compute distances between corresponding elements of the iterable collections
+`a` and `b` according to distance `metric`.
+
+`a` and `b` must have the same number of elements, `r` must be a vector of length
+`length(a) == length(b)`.
+"""
+function colwise(metric::PreMetric, a, b)
+    n = get_common_length(a, b)
+    r = Vector{result_type(metric, a, b)}(undef, n)
+    colwise!(r, metric, a, b)
 end
 
+"""
+    colwise(r::AbstractMatrix, metric::PreMetric,
+              a::AbstractMatrix, b::AbstractMatrix)
+
+Compute distances between each corresponding columns of `a` and `b` according
```

Comment:
Something to note is that these methods are inconsistent with the ones treating `a` and `b` as iterators of columns: a matrix of vectors will be treated differently from a vector of the same vectors. That's probably OK in practice, but that's one of the reasons why I'd like to move to requiring explicitly writing `pairwise(d, eachol(a), eachcol(b))` in the longer term. That way we won't need `dims` anymore.

Reply:
That would be very nice, because it redirects the matrix-based method to the iterator-based method, and one could get rid of the matrix-based ones. The only issue I see is that for the specialized `*Euclidean` (and a few others) distances, where we do need the underlying matrix for performance reason, I don't seem to be able to unwrap it from the `eachcol` generator.

Reply:
Fortunately https://github.com/JuliaLang/julia/pull/32310 should allow us to retrieve the underlying matrix!

Reply:
Yes, I looked at that one a bit today. Will they let it go into v1.6? I wonder how the ecosystem is going to adapt to v1.6 being the new LTS, and how fast packages will really drop 1.6- support. In many cases, there is no hard reason, only soft ones.